### PR TITLE
WIP: Avoid a single recording rule creating many large objects

### PR DESCRIPTION
**Pull request by zecke**
I ran into a problem with a rule like `count_over_time( some_series[Many d:Small s])` that would end up with the kernel OOM killer sacrificing prometheus (besides overcommit being turned off). Taking a pprof/heap profile quickly pointed me to the Point slices and looking at the implementation shows me numSamples as a possible contributor. It's a WIP PR as I am not that familiar with Go (and Prometheus)

Most interestingly the --query.max-samples  only partially helps. It does fail the query but still does bad things to the heap (it seems unable to return memory to the OS, even with GODEBUG=madvdontneed=1). The closest thing I have to a working theory is:

1. Pooled []Point with a large capacity ends up being used by smaller queries. This would keep the slices alive with a lot of memory unused. It seems a best practice for sync.Pool is to have objects of roughly the same memory cost in it. I peeked into Go's fmt implementation and have a patch that discards large slices (larger than the ones used in the benchmark).

1. Interestingly the above does not apply when --query.max-samples kicks in as many slices are not returned/inserted to the pool but I can still see Prometheus taking up most of the physical RAM and never return it to the OS (and then run at risk to get OOM killed). My hypothesis is that Go's memory allocator will try to directly allocate large slices.. and we fragment the heap because we have many of them.
On a conceptual level I think the capacity requested by getPointSlice calls in a query should not exceed maxSamples but that might be rather difficult to implement. Another observation is that numSteps is the upper bound and in some cases we might get a lower bound (Min(numSteps, len(result)) or even get it precisely?

### CODE REVIEW

File: promql/engine.go [hunk 1]
```diff
@@ -1399,15 +1399,36 @@ func (ev *evaluator) vectorSelectorSingle(it *storage.BufferedSeriesIterator, no
...
 var pointPool = sync.Pool{}
 
-func getPointSlice(sz int) []Point {
+func getPointSlice(maxSz int) []Point {
 	p := pointPool.Get()
 	if p != nil {
 		return p.([]Point)
 	}
+
+	// Some queries can blow up the heap (and then run into maxSamples)
```
Comment (brian-brazil) [inline promql/engine.go:1408]: Some queries however will use what they asked for, and giving them a smaller slice will only make them slower as it'll have to be extended many times.

**Review by brian-brazil — COMMENTED** [review]
What the performance/allocation impact of doing this?

This code comes from before subqueries, when num steps could never be bigger than 11k.

**Comment by zecke** [comment]
> What is the performance/allocation impact of doing this?

The benchmark in promql didn't seem conclusive (lot's of noise, some more fast and some more slow) but I picked the threshold in a way that the benchmarks in promql/ do not need to grow the slice.

 
> This code comes from before subqueries, when num steps could never be bigger than 11k.

In my case I have a low retention (days) but this recording rule created a large numSamples (let's say 365d at 1s resolution). And let's assume the query has something like 100 rows in the result set and the total query does not run into the ev.maxSamples limit. Then we have 100 slices with millions of entries capacities (all large allocations) and Go's GC is too poor to deal with that properly.

As a minimum I think we could:

 * Not return large allocations to the pool
 * We could clamp `sz` to `ev.maxSamples` or even `ev.maxSamples - ev.currentSamples`. In the above case the first results would have large allocations but not all rows.


Better: Allocating the exact number but I don't know enough of the data structures to see how difficult it is. I do see things like `biggestLen` but also once we started to iterate from startTs to endTs and create the point slice on first occurrence of a metric's hash.. we know how many samples were not consumed..

WDUT?

**Comment by brian-brazil** [comment]
> The benchmark in promql didn't seem conclusive (lot's of noise, some more fast and some more slow) but I picked the threshold in a way that the benchmarks in promql/ do not need to grow the slice.

We'll need some new benchmarks then.

> Not return large allocations to the pool
> We could clamp sz to ev.maxSamples or even ev.maxSamples - ev.currentSamples.

These would make sense.

> Allocating the exact number but I don't know enough of the data structures to see how difficult it is. 

We don't know the number until the calculation is performed, and copying around data again would be slow and mean more allocations. #6372 is related.

**Comment by zecke** [comment]
> > Allocating the exact number but I don't know enough of the data structures to see how difficult it is.
> 
> We don't know the number until the calculation is performed, and copying around data again would be slow and mean more allocations. #6372 is related.

But is the trade off really in favor of quickly blowing up the plant over running it safely? From my point of view such a query will be slow anyway and growing+copying the slice is unlikely to be the slowest part here. I am totally guessing but neither of us has a benchmark for this degenerated case. ;)

**Comment by brian-brazil** [comment]
This is why benchmarks are needed, extra allocations will affect overall performance due to the additional GC.

**Comment by roidelapluie** [comment]
We have looked at this pull request during our bug scrub.

Because of the lack of benchmark, we have decided to close this issue. Feel free to reopen if needed.

cc @bboreham for awareness.

Thank you for your contribution.

**Comment by zecke** [comment]
@roidelapluie I haven't looked at this part of the code in a while but I think it still makes sense to not return slices with a large capacity to the pool.

**Pull request closed without merging**